File: scripts/download_papers.py
#!/usr/bin/env python3
"""
Download all Dropbox-linked files from content/publications.yml into assets/papers/,
then update the YAML to use local paths. Uses only stdlib (no PyYAML).
Run from project root: python scripts/download_papers.py
"""
import re
import time
import urllib.request
from pathlib import Path
from urllib.parse import urlparse, unquote

PROJECT_ROOT = Path(__file__).resolve().parent.parent
YAML_PATH = PROJECT_ROOT / "content" / "publications.yml"
PAPERS_DIR = PROJECT_ROOT / "assets" / "papers"
DROPBOX_PREFIX = "https://www.dropbox.com/"


def extract_dropbox_urls(yaml_text):
    """Find all lines containing url: https://www.dropbox.com/... and return (full_line, url)."""
    urls = []
    for line in yaml_text.splitlines():
        if "url:" in line and DROPBOX_PREFIX in line:
            # Get the URL part (after "url:")
            m = re.match(r'^(\s*url:\s*)(.+)$', line)
            if m:
                prefix, rest = m.groups()
                url = rest.strip()
                urls.append((line, url))
    return urls


def filename_from_url(url):
    """Last path segment of URL, URL-decoded."""
    parsed = urlparse(url)
    path = parsed.path
    name = path.rstrip("/").split("/")[-1]
    return unquote(name)


def download_file(url, dest_path):
    """Return True if download succeeded."""
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0 (compatible; download-papers/1.0)"})
        with urllib.request.urlopen(req, timeout=60) as resp:
            if resp.status != 200:
                return False
            dest_path.write_bytes(resp.read())
        return True
    except Exception as e:
        print(f"  Failed: {e}")
        return False


def main():
    PAPERS_DIR.mkdir(parents=True, exist_ok=True)

    yaml_text = YAML_PATH.read_text(encoding="utf-8")
    url_lines = extract_dropbox_urls(yaml_text)

    # Build unique filename for each URL (disambiguate duplicates)
    seen_filenames = {}
    url_to_local = {}

    for line, url in url_lines:
        base_name = filename_from_url(url)
        if not base_name:
            continue
        name = base_name
        counter = 0
        while name in seen_filenames and seen_filenames[name] != url:
            counter += 1
            stem, suffix = base_name.rsplit(".", 1) if "." in base_name else (base_name, "")
            if suffix:
                name = f"{stem}_{counter}.{suffix}"
            else:
                name = f"{base_name}_{counter}"
        seen_filenames[name] = url
        dest = PAPERS_DIR / name
        local_path = f"assets/papers/{name}"

        if dest.exists():
            print(f"Already exists: {name}")
            url_to_local[url] = local_path
            continue

        print(f"Downloading: {name} ...")
        if download_file(url, dest):
            url_to_local[url] = local_path
            print(f"  OK -> {local_path}")
        else:
            print(f"  Skip (leave URL unchanged)")
        time.sleep(0.5)

    # Replace URLs in YAML text with local paths
    new_text = yaml_text
    for url, local_path in url_to_local.items():
        # Replace the exact url value (avoid replacing in comments/abstracts)
        old_line = f"url: {url}"
        new_line = f"url: {local_path}"
        if old_line in new_text:
            new_text = new_text.replace(old_line, new_line, 1)

    YAML_PATH.write_text(new_text, encoding="utf-8")
    print(f"\nUpdated {YAML_PATH} with {len(url_to_local)} local paths.")


if __name__ == "__main__":
    main()

File: scripts/download_papers.R
#!/usr/bin/env Rscript
# Download all Dropbox-linked files from content/publications.yml into assets/papers/,
# then update the YAML to use local paths.
# Run from project root: Rscript scripts/download_papers.R

suppressWarnings(suppressPackageStartupMessages(library(yaml)))

args <- commandArgs(trailingOnly = FALSE)
script <- sub("^--file=", "", args[grep("^--file=", args)])
project_root <- if (length(script)) normalizePath(dirname(dirname(script)), winslash = "/", mustWork = FALSE) else getwd()
if (!dir.exists(project_root)) project_root <- getwd()
setwd(project_root)

yaml_path <- file.path(project_root, "content", "publications.yml")
papers_dir <- file.path(project_root, "assets", "papers")
dir.create(papers_dir, recursive = TRUE, showWarnings = FALSE)

pubs <- read_yaml(yaml_path)
url_to_local <- character(0)
seen_names <- character(0)

for (entry in pubs) {
  links <- entry$links %||% list()
  for (i in seq_along(links)) {
    url <- links[[i]]$url
    if (is.null(url) || !startsWith(url, "https://www.dropbox.com/")) next
    path_part <- sub("\\?.*$", "", url)
    parts <- strsplit(path_part, "/", fixed = TRUE)[[1]]
    base_name <- utils::URLdecode(parts[length(parts)])
    if (!nzchar(base_name)) next
    name <- base_name
    k <- 0
    while (name %in% names(seen_names) && seen_names[name] != url) {
      k <- k + 1
      if (grepl("\\.", name)) {
        name <- sub("(.*)\\.([^.]+)$", paste0("\\1_", k, ".\\2"), base_name)
      } else {
        name <- paste0(base_name, "_", k)
      }
    }
    seen_names[name] <- url
    dest <- file.path(papers_dir, name)
    local_path <- file.path("assets", "papers", name)
    local_path <- gsub("\\\\", "/", local_path)
    if (file.exists(dest)) {
      message("Already exists: ", name)
      url_to_local[url] <- local_path
      next
    }
    message("Downloading: ", name, " ...")
    tryCatch({
      utils::download.file(url, dest, mode = "wb", quiet = TRUE)
      if (file.exists(dest) && file.info(dest)$size > 0) {
        url_to_local[url] <- local_path
        message("  OK -> ", local_path)
      }
    }, error = function(e) message("  Failed: ", conditionMessage(e)))
    Sys.sleep(0.5)
  }
}

# Update YAML by text replacement to preserve structure
txt <- readLines(yaml_path, encoding = "UTF-8", warn = FALSE)
for (i in seq_along(url_to_local)) {
  old_url <- names(url_to_local)[i]
  new_path <- url_to_local[i]
  old_line <- paste0("url: ", old_url)
  new_line <- paste0("url: ", new_path)
  txt <- sub(old_line, new_line, txt, fixed = TRUE)
}
writeLines(txt, yaml_path, useBytes = FALSE)
message("\nUpdated ", yaml_path, " with ", length(url_to_local), " local paths.")
